### PR TITLE
feat: add delay_minutes option for cat_change

### LINKED ISSUE #1074: [FR]: Add option for offset number of minutes before a specific category change
### Is your feature request related to a problem? Please elaborate.

No just an enhancement that I think would be worthwhile if people filter certain categories from view.

### Describe the solution you'd like

For the category change setup screen add an option for a time based offset for that category change rule.  This would amount to something like do not change the category until X amount of time has passed since the torrent was completed.

### Does your solution involve any of the following?

- [x] New config option
- [ ] New command option

### Describe alternatives you've considered

Not really any other solutions as this was a straight forward request...

### Who will this benefit?

People who want to delay category changes for torrents that have completed until X amount of time after it completes.

### Additional Information

_No response_

### PR DESCRIPTION
**Pull request by bakerboy448**
## Description

Fixes #1074

Adds an optional `delay_minutes` to `cat_change` entries to delay category changes until N minutes after torrent completion. Uses the qBittorrent `completion_on` timestamp.

### Config format

Supports both simple (existing) and extended formats:

```yaml
cat_change:
  # Simple (backwards compatible):
  CatA.cross-seed: CatA

  # Extended with delay:
  CatB.cross-seed:
    new_cat: CatB
    delay_minutes: 30  # Wait 30 min after completion
```

### How it works

1. Config parsing (`_process_cat_change`) normalizes both formats into `{"new_cat": str, "delay_minutes": int}`
2. `change_categories()` checks `torrent.completion_on` timestamp against `delay_minutes` before changing
3. Torrents that haven't waited long enough are skipped (logged at DEBUG level) and will be picked up on the next run

### Changes
- `modules/config.py`: New `_process_cat_change()` method normalizes config format
- `modules/core/category.py`: `change_categories()` checks delay before changing
- `config/config.yml.sample`: Documented extended format with example
- `web-ui/js/config-schemas/cat_change.js`: Added `delay_minutes` field to schema

Fully backwards compatible — simple string format works identically to before.

## Type of change

- [x] New feature (non-breaking change which adds functionality)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

## Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have added or updated the docstring for new or existing methods
- [x] I have modified this PR to merge to the develop branch

### CODE REVIEW

**Comment by bakerboy448** [comment]
Heads-up — this PR touches `modules/config.py`, which is also edited by #1159 and #1160. Recommended merge order: **#1161 → #1159 → #1160**. #1160 (strict config validation) has a hard logical dependency on the new config keys introduced here (`delay_minutes`, `new_cat` under `cat_change` extended format) — its `KNOWN_*_KEYS` allowlist must include them, otherwise this PR's config would fail strict validation post-#1160. Trivial rebases between this and #1159 (different sections of `config.py`).

**Comment by bakerboy448** [comment]
**Status: ✅ Ready for review**

**Semver suggestion:** **MINOR** bump (4.7.x → 4.8.0). New optional `delay_minutes` field with extended `{new_cat, delay_minutes}` dict form; legacy `old: new` string format remains supported via `_process_cat_change()` normalizer, so no consumer breaks.

**Dependencies on other open PRs:**
- **None** — standalone feature.
- #1160 (strict config validation) currently has no validation surface for `cat_change` keys (the section is a user-defined map, not a fixed-key schema), so this PR doesn't conflict with #1160 regardless of merge order.

**Recommended merge position:** first or second among the 5 open PRs.

**Comment by bakerboy448** [comment]
**Rebased + force-pushed.** Rebased onto current develop. Clean rebase, no conflicts.

**Comment by bakerboy448** [comment]
## Copilot Review Triage — PR #1161

No Copilot review comments found on this PR. Nothing to triage.

**Comment by bakerboy448** [comment]
**Rebased onto develop @ d686923 + force-pushed.** Clean rebase. All CI green (CodeQL, 4× Analyze, label-conflicts, pre-commit.ci).

**Comment by bakerboy448** [comment]
**Copilot suggestions addressed** (commit deded2b)

Both `_process_cat_change()` issues from the Copilot review:

1. **Top-level type check** (`modules/config.py:301`) — `cat_change: []` or `cat_change: ""` now raises `Failed` with a clear message instead of `AttributeError` on `.items()`.
2. **Entry-value type check** (`modules/config.py:316`) — non-string, non-dict entry values (lists, ints, etc.) now raise `Failed` instead of being silently coerced into a literal category name via `str(value)`.

Added `tests/test_config_cat_change.py` — 15 tests covering:
- Invalid top-level types: list, string, integer
- Invalid entry values: list, integer, `None`
- Valid simple form (`{old: "new"}`) and extended form (`{old: {new_cat, delay_minutes}}`)
- Edge cases: missing `new_cat` key, negative/non-numeric `delay_minutes`, float `delay_minutes`

Full suite: 184 → 199 passing locally, no regressions.

**Comment by bakerboy448** [comment]
**Round 2: NEW Copilot bugs from fresh re-review** (commit 0d92c4c)

Copilot's fresh review on deded2b caught 3 NEW bugs I'd missed:

1. **`_process_cat_change()` early-return ran BEFORE type validation** (config.py:301) — falsy non-None values (`cat_change: 0`, `False`, `""`, `[]`) all triggered `if not raw:` and returned `{}` silently. The type check never fired. Fix: validate type FIRST (`raw is None` → `{}`, non-dict → `Failed`), THEN check empty dict.

2. **`bool` satisfied `isinstance(delay, (int, float))`** (config.py:317) — `delay_minutes: True`/`False` silently coerced to `1`/`0` because `bool` is a subclass of `int`. Fix: explicit `isinstance(delay, bool)` guard before the numeric check.

3. **`change_categories()` delay branch was not unit-tested** (category.py:93) — the runtime delay logic had no tests; my prior PR only covered the config-side normalization. Added `tests/core/test_category_delay.py` with 6 tests covering: `delay=0` fires immediately, delay-not-elapsed skips, delay-elapsed fires, `completion_on` sentinel values (`0`, `-1`) skip, boundary (`elapsed == delay`) fires.

Plus added `FakeTorrent.get()` to factories.py to support `torrent.get("completion_on", 0)` calls from production.

Full suite: 205 → 210 passing.

File: modules/config.py [hunk 1]
```diff
@@ -283,6 +283,45 @@ def validate_required_sections(self):
...
+
+        Returns dict mapping old_cat -> {"new_cat": str, "delay_minutes": int}
+        """
+        raw = self.data.get("cat_change")
+        if not raw:
+            return {}
+        if not isinstance(raw, dict):
+            err = f"Config Error: cat_change must be a mapping (dict), got {type(raw).__name__}"
+            self.notify(err, "Config")
+            raise Failed(err)
```
[the remark this answers is not in the document]
Comment (bakerboy448) [reply modules/config.py]: Addressed in subsequent commits — Lines 298-301 validate cat_change is a dict before falsy check; non-dict raises Failed.. Marking thread as resolved.

File: modules/config.py [hunk 2]
```diff
@@ -283,6 +283,38 @@ def validate_required_sections(self):
...
+
+        Returns dict mapping old_cat -> {"new_cat": str, "delay_minutes": int}
+        """
+        raw = self.data.get("cat_change")
+        if not raw:
+            return {}
+        result = {}
+        for old_cat, value in raw.items():
+            if isinstance(value, str):
+                result[old_cat] = {"new_cat": value, "delay_minutes": 0}
```
[the remark this answers is not in the document]
Comment (bakerboy448) [reply modules/config.py]: Addressed in subsequent commits — Lines 298-301 validate raw is a dict with descriptive error message.. Marking thread as resolved.

File: modules/config.py [hunk 3]
```diff
@@ -283,6 +283,47 @@ def validate_required_sections(self):
...
+                new_cat = value.get("new_cat")
+                if not new_cat:
+                    err = f"Config Error: cat_change entry '{old_cat}' is missing required 'new_cat' key"
+                    self.notify(err, "Config")
+                    raise Failed(err)
+                delay = value.get("delay_minutes", 0)
+                if isinstance(delay, bool) or not isinstance(delay, (int, float)) or delay < 0:
+                    err = f"Config Error: cat_change entry '{old_cat}' has invalid delay_minutes: {delay}"
+                    self.notify(err, "Config")
+                    raise Failed(err)
```
[the remark this answers is not in the document]
Comment (bakerboy448) [reply modules/config.py]: Addressed in subsequent commits — Lines 322-328 reject fractional delay_minutes; whole numbers accepted via is_integer() check.. Marking thread as resolved.

File: modules/config.py [hunk 4]
```diff
@@ -283,6 +283,45 @@ def validate_required_sections(self):
...
+                if not new_cat:
+                    err = f"Config Error: cat_change entry '{old_cat}' is missing required 'new_cat' key"
+                    self.notify(err, "Config")
+                    raise Failed(err)
+                delay = value.get("delay_minutes", 0)
+                if not isinstance(delay, (int, float)) or delay < 0:
+                    err = f"Config Error: cat_change entry '{old_cat}' has invalid delay_minutes: {delay}"
+                    self.notify(err, "Config")
+                    raise Failed(err)
+                result[old_cat] = {"new_cat": str(new_cat), "delay_minutes": int(delay)}
```
[the remark this answers is not in the document]
Comment (bakerboy448) [reply modules/config.py]: Addressed in subsequent commits — Line 315 explicit bool check before numeric isinstance check; rejects booleans.. Marking thread as resolved.

File: modules/config.py [hunk 5]
```diff
@@ -283,6 +283,38 @@ def validate_required_sections(self):
...
+            else:
+                result[old_cat] = {"new_cat": str(value), "delay_minutes": 0}
+        return result
```
[the remark this answers is not in the document]
Comment (bakerboy448) [reply modules/config.py]: Addressed in subsequent commits — Lines 330-334 fail fast on unexpected types; no str() coercion.. Marking thread as resolved.

File: modules/core/category.py [hunk 6]
```diff
@@ -79,6 +83,19 @@ def change_categories(self):
...
 
             for torrent in torrent_list:
+                if delay_minutes > 0:
+                    completion_on = torrent.get("completion_on", 0)
+                    # completion_on is -1 or 0 if not yet completed
+                    if completion_on <= 0:
+                        logger.debug(f"Skipping category change for {torrent.name}: torrent has not completed yet")
+                        continue
+                    elapsed_minutes = (now - completion_on) / 60
+                    if elapsed_minutes < delay_minutes:
```
[the remark this answers is not in the document]
Comment (bakerboy448) [reply modules/core/category.py]: Addressed in subsequent commits — tests/core/test_category_delay.py provides comprehensive delay coverage (zero, not-elapsed, elapsed, boundary, not-completed).. Marking thread as resolved.

**Comment by bakerboy448** [comment]
Addressed the Copilot review findings across two commits:

**`ae975c8` (backend)**
- `notify()` before webhooks init → AttributeError masking config errors: moved `_process_cat_change()` to run *after* `process_config_all_webhooks()` wraps `webhooks_factory`.
- `new_cat` accepted any truthy type then `str()`-mangled: now validated as a string.
- delay boundary: `category.py` now compares in **integer seconds** (`elapsed_seconds < delay_minutes * 60`) instead of float minutes.
- test-factory drift: extracted a shared `normalize_cat_change()` pure function (mirrors the `validate_share_limit_action` pattern); `FakeConfig` now uses it so tests track production validation.

**`f9610c0` (web UI)**
- cat_change was the only section on the legacy `dynamic-key-value-list` renderer → extended `{new_cat, delay_minutes}` rendered as `[object Object]` and corrupted config on save. Migrated to the `complex-object` renderer (matching cat/nohardlinks/tracker) with `new_cat` + `delay_minutes` fields; legacy simple-form values are normalized to the object shape on load.

Full suite: 354 passing. Note: the web-ui change has no JS test harness — verified by syntax check + following the existing complex-object pattern; a quick manual smoke in a running UI is worthwhile.

**Review by bobokun — APPROVED** [review]
